src/lib/profiles/data-management/Current/LoggingManagement.cpp

```diff
@@ -436,6 +437,210 @@ void LoggingManagement::DestroyLoggingManagement(void)
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_EventBufferQueueLength), circular_event_buffer->mBuffer.DataLength());
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_FirstEventId), circular_event_buffer->mFirstEventID);
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_LastEventId), circular_event_buffer->mLastEventID);
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_FirstEventTimestamp), circular_event_buffer->mFirstEventTimestamp);
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_LastEventTimestamp), circular_event_buffer->mLastEventTimestamp);
+        SuccessOrExit(err);
+
+#if WEAVE_CONFIG_EVENT_LOGGING_UTC_TIMESTAMPS
+        err = writer.Put(ContextTag(kTag_PersistEvent_FirstEventUTCTimestamp), circular_event_buffer->mFirstEventUTCTimestamp);
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_LastEventUTCTimestamp), circular_event_buffer->mLastEventUTCTimestamp);
+        SuccessOrExit(err);
+
+        err = writer.PutBoolean(ContextTag(kTag_PersistEvent_UTCInitialized), circular_event_buffer->mUTCInitialized);
+        SuccessOrExit(err);
+
+        err = writer.EndContainer(container);
+        SuccessOrExit(err);
+#endif
+    }
+
+    err = writer.EndContainer(rootContainer);
+    SuccessOrExit(err);
+    err = writer.Finalize();
+    SuccessOrExit(err);
+
+    serializedLen = (uint32_t)writer.GetLengthWritten();
+
+exit:
+    Platform::CriticalSectionExit();
+    return err;
+}
+
+/**
+ * @brief Load previously persisted Weave event.
+ *
+ */
+WEAVE_ERROR LoggingManagement::LoadEvent(uint8_t * serializedEvent, uint32_t serializedEventLen)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+
+    TLVReader reader;
+    reader.Init(serializedEvent, serializedEventLen);
+
+    TLVType rootContainer;
+    err = reader.Next(kTLVType_Structure, ProfileTag(kWeaveProfile_Common, kTag_SerializedEvent));
+    SuccessOrExit(err);
+    err = reader.EnterContainer(rootContainer);
+    SuccessOrExit(err);
+
+    for (int importanceType = kImportanceType_First; importanceType < kImportanceType_Last + 1; importanceType++)
+    {
+        CircularEventBuffer * circular_event_buffer = GetImportanceBuffer((ImportanceType)importanceType);
+        TLVType container;
+
+        int eventBufferQueueHeadOffset;
+        size_t eventBufferQueueLength;
+
+        err = reader.Next(kTLVType_Structure, ContextTag(importanceType));
+        SuccessOrExit(err);
+        err = reader.EnterContainer(container);
+        SuccessOrExit(err);
+
+        err = reader.Next(kTLVType_ByteString, ContextTag(kTag_PersistEvent_EventBuffer));
+        SuccessOrExit(err);
+        VerifyOrExit(reader.GetLength() == circular_event_buffer->mBuffer.GetQueueSize(), err = WEAVE_ERROR_INVALID_ARGUMENT);
```

Comment:
This presents a pretty major problem with this serialization format.  If the event buffer sizes aren't an exact match between serializing and deserializing, it is effectively impossible to restore the data.  That means, for example, a firmware update that increased the buffer size would invalidate any persisted data.  That seems very undesirable.

Reply:
I see. Then I guess I should serialize only the event data, join the two part together if the event data hits the physical end address of the ring buffer. And when loading the events back, just check if the buffer size if large enough and set mQueueHead = mQueue. Is this a better way?

Reply:
That sounds more flexible.  One thing to consider is when the buffer is smaller than the data when deserializing.  The ideal implementation would throw away the oldest events in each buffer until the remaining events fit.  For simplicity, you could just throw an error in this case, presuming that buffer sizes only ever grow.